### PR TITLE
refactor(feature_flags): optimize UX and maintenance

### PR DESCRIPTION
**Pull request by heitorlessa**
**Issue #, if available:** #494

## Description of changes:

**Checklist**

 

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [x] Update tests
* [x] Update docs
* [x] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/376ec0a2ac0d2a40e0af5717bef42ff84ca0d1b9/.github/semantic.yml#L2)

## New UX

```python
from typing import List

from aws_lambda_powertools.utilities.feature_flags import FeatureFlags, AppConfigStore

app_config = AppConfigStore(
    environment="test",
    application="powertools",
    name="test_conf_name",
    cache_seconds=600,
)

feature_flags: FeatureFlags = FeatureFlags(store=app_config)

is_my_feature_active: bool = feature_flags.evaluate(name="my_feature", context={}, default=False)

# TODO: This needs reviewing, possibly meant as List[Dict[str, bool]]???
all_features: List[str] = feature_flags.get_enabled_features(context={})
```

**Using envelope argument to pluck feature flags in an arbitrary key from fetched config**

```python
app_config = AppConfigStore(
    environment="test",
    application="powertools",
    name="test_conf_name",
    cache_seconds=600,
    envelope="features"    # feature toggles are under `features` key in fetched config
)

feature_flags: FeatureFlags = FeatureFlags(store=app_config)
```

## Changes

Major changes made during initial doc writing - Due to the size, we're splitting this PR to separate actual docs.

- [x] General
  - [x] ~~Rename `feature_toggles` to `feature_flags`~~
  - [x] ~~Refactor error loggers to contextual exceptions~~
- [x] Improve exceptions
  - [x] ~~Consider an additional fields to improve exception handling such as error type, feature, etc~~
    - [x] ~~String concatenation ended up slowing things down, dropping it~~ 
  - [x] ~~Consider fine grained ConfigurationError instead of a single ConfigurationError~~
  - [x] ~~Rename it to `InvalidSchemaError`~~
- [x] Refactor Schema to ease maintenance
  - [x] ~~Simplify schema by removing `features` key~~
  - [x] ~~Make each rule in `rules` unique by using Dict over List, and remove `rule_name`~~
  - [x] ~~Review long and redundant names such as `"feature": { "feature_default_value"...}`, `"rules": [{"rule_name"...}]`~~
  - [x] ~~Rename `feature_default_Value` to `default`~~
  - [x] ~~Rename `value_when_applies` to `when_match`~~
  - [x] ~~Refactor schema into multiple validators to ease maintenance~~
  - [x] ~~Use module logger over class `logger`~~
  - [x] ~~Create classes for each validation over multiple methods~~
  - [x] ~~Add typing extensions as feature flags and other features become difficult to maintain without things like TypedDict, improved generics for mypy, etc.~~
    	- Note: MyPy doesn't support TypedDic when you use a variable name as a dict key; ignoring it
  - [x] ~~Rename `Action` to `RuleAction` enum~~
  - [x] ~~Suggest what RuleActions are valid when invalid is provided~~
- [x] Refactor ConfigurationStore to `FeatureFlags`
  - [x] `ConfigurationStore` to `FeatureFlags`
  - [x] ~~`rules_context` to `context`~~
  - [x] ~~Rename `get_feature` method to `evaluate`~~
  - [x] ~~Rename `get_all_enabled_feature_toggles` to `get_enabled_features`~~
- [x] Refactor SchemaFetcher to `StoreProvider`
  - [x] `SchemaFetcher` to `StoreProvider`
  - [x] ~~Use `base.py` for interfaces for consistency (e.g. Metrics, Tracer, etc.)~~
  - [x] ~~`AppConfigFetcher` to `AppConfigStore`~~
  - [x] ~~AppConfig construct parameter names for consistency (e.g. `configuration_name` -> `name`, `service` -> `application`)~~
  - [x] ~~Rename `value_if_missing` param to `default` to match Python consistency (e.g. `os.getenv("VAR", default=False)`)~~
  - [x] ~~Rename `get_configuration` method to `get_json_configuration`~~
- [x] Complete docstrings and logging
  - [x] ~~Store section~~
  - [x] ~~FeatureFlags section~~
  - [x] ~~Schema section~~
  - [x] ~~Add schema specification~~
- [x] Test conditional value key being a dictionary
- [x] Load test

**Time allowing**

- [ ] Alternative to testing feature flag private method e.g. `test_is_rule_matched_no_matches`

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws_lambda_powertools/utilities/feature_toggles/schema.py [hunk 1]
```diff
@@ -26,59 +26,69 @@ class SchemaValidator:
     def __init__(self, logger: Logger):
         self._logger = logger
```
Comment (michaelbrewer) [inline aws_lambda_powertools/utilities/feature_toggles/schema.py:27]: @heitorlessa - probably don't need the logger anymore?

Comment (heitorlessa) [reply aws_lambda_powertools/utilities/feature_toggles/schema.py:27]: probably not. I'll look at the schema last, as there are opportunities to make it easier to maintain

**Comment by heitorlessa** [comment]
Rough understanding of the schema as I'm going through before documenting it

```python
## Schema Rules

# `features` MUST be present
# `feature_default_value` MUST be present
# MUST be at least ONE feature
# `rules` MIGHT be present
# `rule_name` must be present
# `rule_default_value` must be BOOL
```

**Comment by heitorlessa** [comment]
It's also quite hard to create features using plain dict/JSON and this manifests in unit testing a function that uses it --- we should create a dataclass or any sort of helper for customers to script the creation of feature flags + unit test. Unsure on what a good solution is yet.

File: aws_lambda_powertools/utilities/feature_flags/__init__.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+from .appconfig import AppConfigStore
+from .base import StoreProvider
+from .exceptions import ConfigurationError
+from .feature_flags import FeatureFlags
+from .schema import ACTION, SchemaValidator
+
+__all__ = [
+    "ConfigurationError",
+    "FeatureFlags",
+    "ACTION",
```
Comment (michaelbrewer) [inline aws_lambda_powertools/utilities/feature_flags/__init__.py:12]: @heitorlessa - Is `ACTION` a bit to generic when imported into your own code? Or is this fine.

**Comment by heitorlessa** [comment]
Initial refactor to Classes to work to separate nested validations: https://gist.github.com/heitorlessa/5a133ef5f08a22a0e046103cb2e40f95

This also allows us to use public method to test specific parts instead of using private methods that will break

File: aws_lambda_powertools/utilities/feature_flags/__init__.py (hunk 2 above)
Comment (heitorlessa) [reply aws_lambda_powertools/utilities/feature_flags/__init__.py]: yep, I've renamed to RuleAction so it also follows `EnumName` "rule"

**Comment by heitorlessa** [comment]
For tomorrow, what do you think of simplifying some of the data structure @am29d @michaelbrewer @risenberg-cyberark @pcolazurdo?

This will make it easier to create a helper class to create and test them (e.g. get all rules, get all features).

## Proposed

```python
features = {
    "my_feature": {
        "default": True,
        "rules": {
            "tenant id equals 345345435": {
                "when_match": False,
                "conditions": [
                    {
                        "action": "EQUALS",
                        "key": "tenant_id",
                        "value": "345345435",
                    }
                ],
            },
        },
    }
}
```


## Current

```python
features = {
    "features": {
        "my_feature": {
            "feature_default_value": True,
            "rules": [
                {
                    "rule_name": "tenant id equals 345345435",
                    "value_when_applies": False,
                    "conditions": [
                        {
                            "action": "EQUALS",
                            "key": "tenant_id",
                            "value": "345345435",
                        }
                    ],
                },
            ],
        }
    },
}
```

**Comment by risenberg-cyberark** [comment]
@heitorlessa i'm fine with when_match but what's the advantage of making the rules a dict instead of list? the current apis support get all matched toggles. why is iterating a dict better than iterating a list?

**Comment by heitorlessa** [comment]
Sure, here's what came to mind:

1) Easier to construct if we're going to offer a helper function to create them

2) Quicker to retrieve rule by name and __contains__ logic, specially when testing

Cosmetically, I personally find it easier to parse when reading

**Comment by michaelbrewer** [comment]
@heitorlessa  looks good to me. Nice and concise.

**Comment by michaelbrewer** [comment]
@heitorlessa  here are the fixes as a diff

```patch
diff --git a/aws_lambda_powertools/utilities/feature_flags/feature_flags.py b/aws_lambda_powertools/utilities/feature_flags/feature_flags.py
index 861f230..3d8100b 100644
--- a/aws_lambda_powertools/utilities/feature_flags/feature_flags.py
+++ b/aws_lambda_powertools/utilities/feature_flags/feature_flags.py
@@ -57,7 +57,7 @@ class FeatureFlags:
                 return False
             # if we got here, all conditions match
             logger.debug(
-                f"rule matched, rule_name={rule_name}, rule_default_value={rule_default_value}, " f"name={feature_name}"
+                f"rule matched, rule_name={rule_name}, rule_default_value={rule_default_value}, name={feature_name}"
             )
             return True
         return False
@@ -160,7 +160,7 @@ class FeatureFlags:
         feature = features.get(name)
         if feature is None:
             logger.debug(
-                f"feature does not appear in configuration, using provided default, " f"name={name}, default={default}"
+                f"feature does not appear in configuration, using provided default, name={name}, default={default}"
             )
             return default
 
diff --git a/aws_lambda_powertools/utilities/feature_flags/schema.py b/aws_lambda_powertools/utilities/feature_flags/schema.py
index cc7216d..09a54b1 100644
--- a/aws_lambda_powertools/utilities/feature_flags/schema.py
+++ b/aws_lambda_powertools/utilities/feature_flags/schema.py
@@ -38,12 +38,9 @@ class SchemaValidator:
 
 
 class FeaturesValidator(BaseValidator):
-    def __init__(self, schema):
+    def __init__(self, schema: Dict):
         self.schema = schema
-        self.features: Optional[Dict[str, Dict]] = None
-
-        if isinstance(self.schema, dict):
-            self.features = self.schema.get(FEATURES_KEY)
+        self.features: Optional[Dict[str, Dict]] = self.schema.get(FEATURES_KEY)
 
     def validate(self):
         if not isinstance(self.features, dict):
@@ -71,13 +68,13 @@ class RulesValidator(BaseValidator):
         self.rules: Optional[List[Dict]] = self.feature.get(RULES_KEY)
 
     def validate(self):
-        if not isinstance(self.rules, list):
-            raise ConfigurationError(f"Feature rules is not a list, feature_name={self.feature_name}")
-
         if not self.rules:
             logger.debug("Rules are empty, ignoring validation")
             return
 
+        if not isinstance(self.rules, list):
+            raise ConfigurationError(f"Feature rules is not a list, feature_name={self.feature_name}")
+
         for rule in self.rules:
             self.validate_rule(rule, self.feature)
             conditions = ConditionsValidator(rule=rule, rule_name=rule.get(RULE_NAME_KEY))
diff --git a/tests/functional/feature_toggles/test_feature_toggles.py b/tests/functional/feature_toggles/test_feature_toggles.py
index 295e762..88cf455 100644
--- a/tests/functional/feature_toggles/test_feature_toggles.py
+++ b/tests/functional/feature_toggles/test_feature_toggles.py
@@ -279,7 +279,7 @@ def test_toggles_match_rule_with_contains_action(mocker, config):
                 "feature_default_value": False,
                 "rules": [
                     {
-                        "rule_name": "tenant id is contained in [6,2] ",
+                        "rule_name": "tenant id is contained in [6, 2]",
                         "value_when_applies": expected_value,
                         "conditions": [
                             {
@@ -306,7 +306,7 @@ def test_toggles_no_match_rule_with_contains_action(mocker, config):
                 "feature_default_value": expected_value,
                 "rules": [
                     {
-                        "rule_name": "tenant id is contained in [6,2] ",
+                        "rule_name": "tenant id is contained in [8, 2]",
                         "value_when_applies": True,
                         "conditions": [
                             {
@@ -333,7 +333,7 @@ def test_multiple_features_enabled(mocker, config):
                 "feature_default_value": False,
                 "rules": [
                     {
-                        "rule_name": "tenant id is contained in [6,2] ",
+                        "rule_name": "tenant id is contained in [6, 2]",
                         "value_when_applies": True,
                         "conditions": [
                             {
@@ -366,7 +366,7 @@ def test_multiple_features_only_some_enabled(mocker, config):
                 "feature_default_value": False,
                 "rules": [
                     {
-                        "rule_name": "tenant id is contained in [6,2] ",
+                        "rule_name": "tenant id is contained in [6, 2]",
                         "value_when_applies": True,
                         "conditions": [
                             {
diff --git a/tests/functional/feature_toggles/test_schema_validation.py b/tests/functional/feature_toggles/test_schema_validation.py
index 8329dd6..6495a00 100644
--- a/tests/functional/feature_toggles/test_schema_validation.py
+++ b/tests/functional/feature_toggles/test_schema_validation.py
@@ -33,11 +33,13 @@ def test_invalid_features_dict():
 
     schema = []
     # invalid type
+    validator = SchemaValidator(schema)
     with pytest.raises(ConfigurationError):
         validator.validate()
 
     # invalid features key
     schema = {FEATURES_KEY: []}
+    validator = SchemaValidator(schema)
     with pytest.raises(ConfigurationError):
         validator.validate()
 
@@ -57,33 +59,38 @@ def test_invalid_feature_dict():
 
     # empty feature dict
     schema = {FEATURES_KEY: {"my_feature": {}}}
+    validator = SchemaValidator(schema)
     with pytest.raises(ConfigurationError):
         validator.validate()
 
     # invalid FEATURE_DEFAULT_VAL_KEY type, not boolean
     schema = {FEATURES_KEY: {"my_feature": {FEATURE_DEFAULT_VAL_KEY: "False"}}}
+    validator = SchemaValidator(schema)
     with pytest.raises(ConfigurationError):
         validator.validate()
 
     # invalid FEATURE_DEFAULT_VAL_KEY type, not boolean #2
     schema = {FEATURES_KEY: {"my_feature": {FEATURE_DEFAULT_VAL_KEY: 5}}}
+    validator = SchemaValidator(schema)
     with pytest.raises(ConfigurationError):
         validator.validate()
 
     # invalid rules type, not list
     schema = {FEATURES_KEY: {"my_feature": {FEATURE_DEFAULT_VAL_KEY: False, RULES_KEY: "4"}}}
+    validator = SchemaValidator(schema)
     with pytest.raises(ConfigurationError):
         validator.validate()
 
 
 def test_valid_feature_dict():
-    # no rules list at all
-    schema = {FEATURES_KEY: {"my_feature": {FEATURE_DEFAULT_VAL_KEY: False}}}
+    # empty rules list
+    schema = {FEATURES_KEY: {"my_feature": {FEATURE_DEFAULT_VAL_KEY: False, RULES_KEY: []}}}
     validator = SchemaValidator(schema)
     validator.validate()
 
-    # empty rules list
-    schema = {FEATURES_KEY: {"my_feature": {FEATURE_DEFAULT_VAL_KEY: False, RULES_KEY: []}}}
+    # no rules list at all
+    schema = {FEATURES_KEY: {"my_feature": {FEATURE_DEFAULT_VAL_KEY: False}}}
+    validator = SchemaValidator(schema)
     validator.validate()
 
 
@@ -118,6 +125,7 @@ def test_invalid_rule():
             }
         }
     }
+    validator = SchemaValidator(schema)
     with pytest.raises(ConfigurationError):
         validator.validate()
 
@@ -135,6 +143,7 @@ def test_invalid_rule():
             }
         }
     }
+    validator = SchemaValidator(schema)
     with pytest.raises(ConfigurationError):
         validator.validate()
 
@@ -149,6 +158,7 @@ def test_invalid_rule():
             }
         }
     }
+    validator = SchemaValidator(schema)
     with pytest.raises(ConfigurationError):
         validator.validate()
 
@@ -163,6 +173,7 @@ def test_invalid_rule():
             }
         }
     }
+    validator = SchemaValidator(schema)
     with pytest.raises(ConfigurationError):
         validator.validate()
 
@@ -202,6 +213,7 @@ def test_invalid_condition():
             }
         }
     }
+    validator = SchemaValidator(schema)
     with pytest.raises(ConfigurationError):
         validator.validate()
 
@@ -224,6 +236,7 @@ def test_invalid_condition():
             }
         }
     }
+    validator = SchemaValidator(schema)
     with pytest.raises(ConfigurationError):
         validator.validate()
```

**Comment by michaelbrewer** [comment]
@heitorlessa - here is my fixes as a PR : https://github.com/heitorlessa/aws-lambda-powertools-python/pull/36

File: aws_lambda_powertools/utilities/feature_flags/feature_flags.py [hunk 3]
```diff
@@ -132,50 +148,48 @@ def get_feature_toggle(
...
+
         # look for first rule match
-        logger.debug(
-            f"looking for rule match,  feature_name={feature_name}, feature_default_value={feature_default_value}"
-        )
+        logger.debug(f"looking for rule match,  name={name}, feature_default_value={feature_default_value}")
         return self._handle_rules(
-            feature_name=feature_name,
-            rules_context=rules_context,
+            feature_name=name,
```
Comment (michaelbrewer) [inline aws_lambda_powertools/utilities/feature_flags/feature_flags.py:180]: @heitorlessa seems like some times it is `name` and other times it is `feature_name`.

Comment (heitorlessa) [reply aws_lambda_powertools/utilities/feature_flags/feature_flags.py:180]: Yep I'm still making a list of variables name to go after.

I hope I can get to the point where these can be abstracted by a higher level class so we reduce the amount of variable propagation

Comment (heitorlessa) [reply aws_lambda_powertools/utilities/feature_flags/feature_flags.py:180]: Or better yet, handle exceptions higher up

**Comment by risenberg-cyberark** [comment]
@heitorlessa  @michaelbrewer, Hi, i think it's a mistake to remove the 'features' key.  It was put there on purpose as a way to distinguish between  features that are related to appconfig and regular configuration that you want to store: any other dict/list etc.
It will also make it  much harder to validate the schema because you might encounter regular conf, nto just toggles. I use my app config nto just for toggles but also for regular config - log level etc.

**Comment by heitorlessa** [comment]
Hi @risenberg-cyberark replying it here so it's documented -- Here's how we can solve this, just like we do for validator, parse, etc.

```python
app_config = AppConfigStore(
    environment="test",
    application="powertools",
    name="test_conf_name",
    cache_seconds=600,
    envelope="features"    # JMESPath expression
)

feature_flags: FeatureFlags = FeatureFlags(store=app_config)
```

By providing an `envelope` we can pluck wherever you find it best to store in your config. This also allows us to have a simpler schema and easier to maintain overall for both of us (end users and maintainers).

I'll work on that now

**Comment by heitorlessa** [comment]
Done @risenberg-cyberark, by passing `envelope="features"` or any JMESPath including our built-in JMESPath functions to deserialize JSON strings etc. will work here 

```python
app_config = AppConfigStore(
    environment="test",
    application="powertools",
    name="test_conf_name",
    cache_seconds=600,
    envelope="features"    # Any JMESPath expression will pluck the features after config is fetched
)

feature_flags: FeatureFlags = FeatureFlags(store=app_config)
```

cc @am29d

**Comment by michaelbrewer** [comment]
@risenberg-cyberark @heitorlessa - this is looking good. Are we still labeling this as beta? In case there are some small refactoring later.

File: aws_lambda_powertools/utilities/feature_flags/feature_flags.py (hunk 3 above)
Comment (heitorlessa) [reply aws_lambda_powertools/utilities/feature_flags/feature_flags.py]: addressed.

**Comment by heitorlessa** [comment]
> @risenberg-cyberark @heitorlessa - this is looking good. Are we still labeling this as beta? In case there are some small refactoring later.

After this refactoring, I'd go with not labelling as beta, unless something major shows up during the load testing.

After launch, we could also consider a generic `AnalyticsProvider` with a `emit` method. This will enrich customers' understanding of their customers and what features might have had enough data points (used, not used).

Two things pending then we merge: **1/** Exception Handling: InvalidSchema over ConfigurationError, and catch Store exceptions and raise accordingly, **2** Load testing: I'm slightly concerned about performance so need to check whether Store cache is sufficient or if we need to reduce the amount of `get_configuration` calls that happen all the time -- could be 1ms for all validation logic, dunno.

**Comment by heitorlessa** [comment]
Exceptions are now improved. Just need an extra test for condition values that are not just string to be sure we got it right.

I'll load test tomorrow, merge it, and get on with the docs with @am29d to launch it by Friday EOD/Monday in the worst case scenario (life happens!)

**Comment by michaelbrewer** [comment]
@heitorlessa minor patch update

```patch
diff --git a/aws_lambda_powertools/utilities/feature_flags/__init__.py b/aws_lambda_powertools/utilities/feature_flags/__init__.py
index 1edbea1..5514a50 100644
--- a/aws_lambda_powertools/utilities/feature_flags/__init__.py
+++ b/aws_lambda_powertools/utilities/feature_flags/__init__.py
@@ -1,5 +1,4 @@
-"""Advanced feature toggles utility
-"""
+"""Advanced feature toggles utility"""
 from .appconfig import AppConfigStore
 from .base import StoreProvider
 from .exceptions import ConfigurationStoreError
diff --git a/aws_lambda_powertools/utilities/feature_flags/appconfig.py b/aws_lambda_powertools/utilities/feature_flags/appconfig.py
index ea3f5dd..6dd6292 100644
--- a/aws_lambda_powertools/utilities/feature_flags/appconfig.py
+++ b/aws_lambda_powertools/utilities/feature_flags/appconfig.py
@@ -69,10 +69,13 @@ class AppConfigStore(StoreProvider):
         """
         try:
             # parse result conf as JSON, keep in cache for self.max_age seconds
-            config = self._conf_store.get(
-                name=self.name,
-                transform=TRANSFORM_TYPE,
-                max_age=self.cache_seconds,
+            config = cast(
+                dict,
+                self._conf_store.get(
+                    name=self.name,
+                    transform=TRANSFORM_TYPE,
+                    max_age=self.cache_seconds,
+                ),
             )
 
             if self.envelope:
@@ -80,6 +83,6 @@ class AppConfigStore(StoreProvider):
                     data=config, envelope=self.envelope, jmespath_options=self.jmespath_options
                 )
 
-            return cast(dict, config)
+            return config
         except (GetParameterError, TransformParameterError) as exc:
             raise ConfigurationStoreError("Unable to get AWS AppConfig configuration file") from exc
```

**Comment by heitorlessa** [comment]
When doing the load testing I found a silent bug - We're swallowing all types of exceptions, including AccessDenied when fetching config from AppConfig.

I'll work on a generic `StoreClientError` and ensure these and future exceptions from any providers are bubbled up correctly

**Comment by heitorlessa** [comment]
First load test with a single feature being evaluated - 10 concurrent reqs, 2K RPS

```
All virtual users finished
Summary report @ 22:27:28(+0200) 2021-08-04
  Scenarios launched:  10
  Scenarios completed: 10
  Requests completed:  2000
  Mean response/sec: 161.03
  Response time (msec):
    min: 30
    max: 1122
    median: 49
    p95: 67
    p99: 155.5
  Scenario counts:
    0: 10 (100%)
  Codes:
    200: 2000
```

**Comment by heitorlessa** [comment]
Happy with the performance and we can merge now. This is now maintainable, it has docstrings to aid customers using it as well as maintainers, I've also created a schema specification, and it's easier to extend the feature schema, plus more compact and concise.

After launch, when time allows, we can easily add a AnalyticsProvider where one could bring their own to track who's using features that match conditions.

Thanks a lot @michaelbrewer for the help in the refactoring too. 

@am29d we're now ready to move onto the docs, then release it.

**Comment by pcolazurdo** [comment]
> if we need to reduce the amount of `get_configuration` calls that happen all the time -- could be 1ms for all validation logic, dunno.

We discussed about using memoize to speed up this, but @risenberg-cyberark pointed out that the rules could be really large in some cases and the memory used may have the opposite effect. This may be something to revisit if performance test aren't great.

**Pull request merged**